src/runtime-rs/crates/runtimes/src/shim_metrics.rs

```diff
@@ -0,0 +1,245 @@
+// Copyright 2021-2022 Kata Contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+extern crate procfs;
+
+use anyhow::{anyhow, Result};
+use prometheus::{Encoder, Gauge, GaugeVec, Opts, Registry, TextEncoder};
+use slog::warn;
+use std::sync::Once;
+
+const NAMESPACE_KATA_SHIM: &str = "kata_shim";
+
+// Convenience macro to obtain the scope logger
+macro_rules! sl {
+    () => {
+        slog_scope::logger().new(o!("subsystem" => "metrics"))
+    };
+}
+
+lazy_static! {
+
+    static ref INIT_REGISTER: Once = Once::new();
+
+    // custom registry
+    static ref REGISTRY: Registry = Registry::new();
+
+    // shim metrics
+    static ref SHIM_THREADS: Gauge = Gauge::new(format!("{}_{}", NAMESPACE_KATA_SHIM, "threads"),"Kata containerd shim v2 process threads.").unwrap();
+
+    static ref SHIM_PROC_STATUS: GaugeVec =
+        GaugeVec::new(Opts::new(format!("{}_{}",NAMESPACE_KATA_SHIM,"proc_status"), "Kata containerd shim v2 process status."), &["item"]).unwrap();
+
+    static ref SHIM_PROC_STAT: GaugeVec =GaugeVec::new(Opts::new(format!("{}_{}",NAMESPACE_KATA_SHIM,"proc_stat"), "Kata containerd shim v2 process statistics."), &["item"]).unwrap();
+
+    static ref SHIM_NETDEV: GaugeVec =GaugeVec::new(Opts::new(format!("{}_{}",NAMESPACE_KATA_SHIM,"netdev"), "Kata containerd shim v2 network devices statistics."), &["interface", "item"]).unwrap();
+
+    static ref SHIM_IO_STAT: GaugeVec =GaugeVec::new(Opts::new(format!("{}_{}",NAMESPACE_KATA_SHIM,"io_stat"), "Kata containerd shim v2 process IO statistics."), &["item"]).unwrap();
+
+    static ref SHIM_OPEN_FDS: Gauge = Gauge::new(format!("{}_{}", NAMESPACE_KATA_SHIM, "fds"), "Kata containerd shim v2 open FDs.").unwrap();
+
+    // TODO:
+    // RPC_DURATIONS_HISTOGRAM & SHIM_POD_OVERHEAD_CPU & SHIM_POD_OVERHEAD_MEMORY
+
+    // static ref RPC_DURATIONS_HISTOGRAM : HistogramVec = HistogramVec::new(HistogramOpts::new(format!("{}_{}", NAMESPACE_KATA_SHIM, "rpc_durations_histogram_milliseconds"),"RPC latency distributions."),&["action"]).unwrap();
+
+    // static ref SHIM_POD_OVERHEAD_CPU: Gauge = Gauge::new(format!("{}_{}", NAMESPACE_KATA_SHIM, "pod_overhead_cpu"), "Kata Pod overhead for CPU resources(percent).").unwrap();
+
+    // static ref SHIM_POD_OVERHEAD_MEMORY: Gauge = Gauge::new(format!("{}_{}", NAMESPACE_KATA_SHIM, "pod_overhead_memory_in_bytes"), "Kata Pod overhead for memory resources(bytes).").unwrap();
+}
+
+pub async fn get_shim_metrics() -> Result<String> {
```

Comment:
Is there an opportunity for us here to specify the categories of metrics that are collected once, rather than collecting all metrics at once?

Reply:
Perhaps you mean the code in [src/runtime-rs/crates/runtimes/src/shim_mgmt/handlers.rs](https://github.com/kata-containers/kata-containers/pull/5797/files/e591453c12bcc556594532be67b0d1b3cf84ae71#diff-2c6f65e82b6142d3d77d8599182a52257b90f6079e3d7585f29baf13bce615deR106-R117)?

```
async fn metrics_url_handler(
     sandbox: Arc<dyn Sandbox>,
     _req: Request<Body>,
 ) -> Result<Response<Body>> {
     // get metrics from agent, hypervisor, and shim in parallel
     let agent_sandbox = sandbox.clone();
     let hypervisor_sandbox = sandbox.clone();

     let handle_agent =
         tokio::task::spawn(async move { agent_sandbox.agent_metrics().await.unwrap() });
```

Reply:
This is one thing, we may only want metrics for part of all components(shim/hypervisor/agent...). On the other, for several metrics(shim_threads/shim_proc_status/shim_proc_stat...) of one component, we may not care about all of them.

Reply:
Prometheus only supports collecting all metrics at once by `let metric_families = REGISTRY.gather();`.
To achieve the function you proposed, we may need to pass a vector of metric names when calling get_*_metrics, and then perform string filtering on the collected results.
I'm a bit confused it's worth doing this. Maybe this can become an advanced feature in monitor-go/monitor-rs?